### PR TITLE
Extend comment-based testing framework with quick fix support

### LINKED ISSUE #3047: Extend comment-based testing framework with quick fix support

### PR DESCRIPTION
**Pull request by victor-diez-sonarsource**
Fixes #3047

### CODE REVIEW

**Comment by ilia-kebets-sonarsource** [comment]
@victor-diez-sonarsource Could you update the DEV.md doc: https://github.com/SonarSource/SonarJS/blob/master/docs/DEV.md#testing-the-rule

File: eslint-bridge/tests/tools/testers/comment-based/helpers/quickfixes.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        : secondColumnType === 'sc'
+        ? +secondColumnValue
+        : undefined;
+    quickfix.end =
+      firstColumnType === 'ec'
+        ? +firstColumnValue
+        : secondColumnType === 'ec'
+        ? +secondColumnValue
+        : undefined;
+    quickfix.fix = fix;
```
Comment (yassin-kammoun-sonarsource) [inline eslint-bridge/tests/tools/testers/comment-based/helpers/quickfixes.ts:99]: Should we add an assertion that checks that start is not greater than end? Not sure what's going to happen when we slice. In any case, we need a unit test for this corner case.

**Review by yassin-kammoun-sonarsource — CHANGES_REQUESTED** [review]
Looks already nice, but we need further tests to validate the changes.`framework.test.ts` is the place to go to add further tests.

Also, we can entirely migrate `inverted-assertion-arguments.test.ts` and `no-exclusive-tests.test.ts`. That should be a good opportunity to validate the work on non-decorated rules.

Following @vilchik-elena's recent guidelines for commits and pull requests, please rename the PR title. I think you can just reuse the ticket's title.

Correct me if I am wrong, but I think we are currently not supporting multiple suggestions for a single issue like this: 

https://github.com/SonarSource/SonarJS/blob/b866d18e80746b6c4c93e1a6afa2bca4acf883ca/eslint-bridge/tests/linting/eslint/rules/decorators/use-isnan-decorator.test.ts#L33-L47

I think it's fine as a first step, but we should create a ticket to track this missing feature unless you really want to implement it now.

Finally, as @ilia-kebets-sonarsource mentioned, we need to document the quickfix support in the DEV documentation so that we know how to use it and what we are actually supporting.

File: eslint-bridge/tests/tools/testers/comment-based/launcher.test.ts [hunk 2]
```diff
@@ -134,9 +109,10 @@ function languageFromFilePath(filePath: string): Language {
  * Loading the above parseForESLint() function.
  */
 const ruleTester = new RuleTester({ parser: __filename });
-const externalRules = {
-  ...Object.fromEntries(new Linter().getRules()),
-  ...reactESLintRules,
+const externalRules = decorateExternalRules({
```
Comment (yassin-kammoun-sonarsource) [inline eslint-bridge/tests/tools/testers/comment-based/launcher.test.ts:112]: Never thought of doing that. Great idea!

File: eslint-bridge/tests/tools/testers/comment-based/launcher.test.ts [hunk 3]
```diff
@@ -123,7 +98,7 @@ export function parseForESLint(
  */
 function languageFromFilePath(filePath: string): Language {
   const { ext } = path.parse(filePath);
-  if (['ts', 'tsx'].includes(ext)) {
+  if (['.ts', '.tsx'].includes(ext)) {
```
Comment (yassin-kammoun-sonarsource) [inline eslint-bridge/tests/tools/testers/comment-based/launcher.test.ts:101]: Ouch :(

File: eslint-bridge/tests/tools/testers/comment-based/launcher.test.ts [hunk 4]
```diff
@@ -33,61 +33,36 @@
...
-      files.forEach(filename => {
-        const ruleModule = rule in internal ? internal[rule] : decorators[rule](external[rule]);
-        const code = fs.readFileSync(filename, { encoding: 'utf8' });
-        const errors = extractExpectations(code, hasSonarRuntimeOption(ruleModule, rule));
+function runRuleTests(rules: Record<string, Rule.RuleModule>, ruleTester: RuleTester) {
+  const tests = fs.readdirSync(fixtures);
+  for (const test of tests) {
+    const filename = path.join(fixtures, test);
+    const { ext, name } = path.parse(filename);
+    const rule = name.toLowerCase();
```
Comment (yassin-kammoun-sonarsource) [inline eslint-bridge/tests/tools/testers/comment-based/launcher.test.ts:54]: I don't think we have a rule that enforces lower-case. Did you encounter a case where this is needed?

File: docs/DEV.md [hunk 5]
```diff
@@ -110,11 +110,55 @@ These tests are located in `eslint-bridge/tests/linting/eslint/rules/comment-bas
 
 ```javascript
 some.clean.code();
-some.faulty.code(); // Noncompliant {{Optional message to assert}}
+some.faulty.code(); // Noncompliant N [[qf1,qf2,...]] {{Optional message to assert}}
 //   ^^^^^^
+// fix@qf1 {{Optional suggestion description}}
+// edit@qf1 [[sc=1;ec=5]] {{text to replace line from [sc] column to [ec] column}}
 ```
+The issue primary location (`// ^^^^`), issue message, issue number (`N`) and quickfixes are optional.
```
Comment (yassin-kammoun-sonarsource) [inline docs/DEV.md:118]:
```suggestion
The issue primary location (`// ^^^^`), issue message, issue number (`N`) and quick fixes are optional.
```

File: docs/DEV.md [hunk 6]
```diff
@@ -110,11 +110,55 @@ These tests are located in `eslint-bridge/tests/linting/eslint/rules/comment-bas
...
 
-The issue primary location (`// ^^^^`) and issue message are optional.
+`N` is an integer defining the amount of issues will be reported in the line.
```
Comment (yassin-kammoun-sonarsource) [inline docs/DEV.md:120]:
```suggestion
`N` is an integer defining the number of issues that will be reported in the line.
```

File: docs/DEV.md [hunk 7]
```diff
@@ -110,11 +110,55 @@ These tests are located in `eslint-bridge/tests/linting/eslint/rules/comment-bas
...
+
+Only one of the methods (`{{messageN}}+` OR `N`) to define the expected issues can be used in a `Noncompliant` line.
+
+If no `N` nor messages are provided, the engine will expect one issue to be raised. Meaning, `//Noncompliant` is equivalent to `//Noncompliant 1`.
```
Comment (yassin-kammoun-sonarsource) [inline docs/DEV.md:124]:
```suggestion
If neither `N` nor messages are provided, the engine will expect one issue to be raised. Meaning, `//Noncompliant` is equivalent to `//Noncompliant 1`.
```

File: docs/DEV.md [hunk 8]
```diff
@@ -110,11 +110,55 @@ These tests are located in `eslint-bridge/tests/linting/eslint/rules/comment-bas
...
+
+The `fix@` comment of a quickfix providing the suggestion description is optional. On the `edit@` comment the start column and end column are also optional, meaning this syntax can be used too:
```
Comment (yassin-kammoun-sonarsource) [inline docs/DEV.md:126]:
```suggestion
The `fix@` comment of a quick fix providing the suggestion description is optional. On the `edit@` comment, the start column and end column are also optional, meaning this syntax can be used too:
```

File: docs/DEV.md [hunk 9]
```diff
@@ -110,11 +110,55 @@ These tests are located in `eslint-bridge/tests/linting/eslint/rules/comment-bas
...
+
+`// edit@qf1 {{text to replace the whole line -do not include //Noncompliant comment- }}`
+
+Alternatively, one can conveniently use only `sc` or `ec`.
+
+Please note the length of the list of quickfixes cannot surpass the amount of issues declared by `N` or the number of expected messages unless their matching issue is reassigned (see below).
```
Comment (yassin-kammoun-sonarsource) [inline docs/DEV.md:132]:
```suggestion
Note that the length of the list of quick fixes cannot surpass the number of issues declared by `N` or the number of expected messages unless their matching issue is reassigned (see below). 
```

File: docs/DEV.md [hunk 10]
```diff
@@ -110,11 +110,55 @@ These tests are located in `eslint-bridge/tests/linting/eslint/rules/comment-bas
...
+
+`// edit@qf1 {{text to replace the whole line -do not include //Noncompliant comment- }}`
+
+Alternatively, one can conveniently use only `sc` or `ec`.
+
+Please note the length of the list of quickfixes cannot surpass the amount of issues declared by `N` or the number of expected messages unless their matching issue is reassigned (see below). 
+
+Quickfixes IDs can be any `string`, they don't have to follow the `qfN` convention. The order of the list is important, as they will be assigned to the message in the matching position. If one provides 3 messages and 2 quickfixes which are not to be matched against first and second message, there are two options:
+
+* A *dummy* quickfix can be used as placeholder:
```
Comment (yassin-kammoun-sonarsource) [inline docs/DEV.md:136]:
```suggestion
* A *dummy* quick fix can be used as placeholder:
```

File: docs/DEV.md [hunk 11]
```diff
@@ -110,11 +110,55 @@ These tests are located in `eslint-bridge/tests/linting/eslint/rules/comment-bas
...
+
+```javascript
+some.faulty.code(); // Noncompliant [[qf1,qf2,qf3]] {{message1}} {{message2}} {{message3}}
+// edit@qf1 {{fix for message1}}
+// edit@qf3 {{fix for message3}}
+// qf2 is declared but never used --> ignored by the engine
+```
+
+* Explicitly set the index (0-based) of the message to which the quickfix refers to with the grammar `=index` next to the quickfix ID:
```
Comment (yassin-kammoun-sonarsource) [inline docs/DEV.md:145]:
```suggestion
* Explicitly set the index (0-based) of the message to which the quick fix refers to with the syntax `=index` next to the quick fix ID:
```

File: docs/DEV.md [hunk 12]
```diff
@@ -110,11 +110,55 @@ These tests are located in `eslint-bridge/tests/linting/eslint/rules/comment-bas
...
+
+Quickfixes IDs can be any `string`, they don't have to follow the `qfN` convention. The order of the list is important, as they will be assigned to the message in the matching position. If one provides 3 messages and 2 quickfixes which are not to be matched against first and second message, there are two options:
+
+* A *dummy* quickfix can be used as placeholder:
+
+```javascript
+some.faulty.code(); // Noncompliant [[qf1,qf2,qf3]] {{message1}} {{message2}} {{message3}}
+// edit@qf1 {{fix for message1}}
+// edit@qf3 {{fix for message3}}
+// qf2 is declared but never used --> ignored by the engine
```
Comment (yassin-kammoun-sonarsource) [inline docs/DEV.md:142]: I think enforcing a convention for a quick fix identifier at some point would be nice. If the digit after `qf` is mandatory, we can identify to which issue it refers and avoid declaring a quick fix that is ignored. Maybe worth creating a ticket to improve that. WDYT?

File: eslint-bridge/tests/tools/testers/comment-based/framework.ts [hunk 13]
```diff
@@ -50,39 +51,120 @@ export function extractExpectations(
...
+    }
+  }
+}
+
+interface Suggestions {
+  suggestions?: RuleTester.SuggestionOutput[];
+}
+
+/**
+ * Applies quickfix edits to a sourcecode line. The fixed line will be formed by a
```
Comment (yassin-kammoun-sonarsource) [inline eslint-bridge/tests/tools/testers/comment-based/framework.ts:120]:
```suggestion
 * Applies quick fix edits to a source code line. The fixed line will be formed by a
```

File: eslint-bridge/tests/tools/testers/comment-based/framework.ts [hunk 14]
```diff
@@ -50,39 +51,120 @@ export function extractExpectations(
...
+          if (containsNC >= 0) {
+            appendAfterFix = line.slice(containsNC);
+          }
+        } else {
+          if (end < start) {
+            throw new Error(`End column cannot be lower than start position ${end} < ${start}`);
+          }
+          if (containsNC >= 0 && end > containsNC) {
+            throw new Error(
+              `End column cannot be in //Noncompliant comment ${end} > ${containsNC}`,
```
Comment (yassin-kammoun-sonarsource) [inline eslint-bridge/tests/tools/testers/comment-based/framework.ts:154]:
```suggestion
              `End column cannot be in // Noncompliant comment ${end} > ${containsNC}`,
```

File: eslint-bridge/tests/tools/testers/comment-based/helpers/quickfixes.ts [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+    // replacement string, ex: {{foo}}
+    ' *(?:\\{\\{(?<fix>.*?)\\}\\}(?!\\}))?' +
+    ' *(?:\r(\n?)|\n)?',
+);
+
+export class QuickFix {
+  public start: number | undefined = undefined;
+  public end: number | undefined = undefined;
+  public description: string | undefined = undefined;
+  public fix: string | undefined = undefined;
```
Comment (yassin-kammoun-sonarsource) [inline eslint-bridge/tests/tools/testers/comment-based/helpers/quickfixes.ts:54]: I think those initializations are redundant.

File: eslint-bridge/tests/tools/testers/comment-based/framework.test.ts [hunk 16]
```diff
@@ -201,4 +201,101 @@ describe('Comment-based Testing Framework', () => {
     const result = await assertions('parsing.js');
     expect(result).toEqual([{ line: 1 }]);
   });
+
```
Comment (yassin-kammoun-sonarsource) [inline eslint-bridge/tests/tools/testers/comment-based/framework.test.ts:204]: Let's try to be consistent with the test above and define fixture files rather than inline code for both code and expected output.

**Review by yassin-kammoun-sonarsource — APPROVED** [review]
Great work! That required much more work than expected, but we support plenty of cases. I think there are a few things that we can improve, but we can postpone them and create improvement tickets.

I left minor suggestions that you are free to ignore if you don't find them relevant.

File: docs/DEV.md (hunk 12 above)
Comment (victor-diez-sonarsource) [reply docs/DEV.md:142]: problem is that right now quick fix ids have to be unique file-wise. So a test cannot have multiple qf1 declared, if you want to have always the 1 pointing to the first issue of each issueLine.

File: eslint-bridge/tests/tools/testers/comment-based/launcher.test.ts (hunk 4 above)
Comment (victor-diez-sonarsource) [reply eslint-bridge/tests/tools/testers/comment-based/launcher.test.ts:54]: just in case we find a filename with any uppercase letter, to properly match against rule key
